docs/fn-2.0/README.md

```diff
@@ -0,0 +1,416 @@
+
+### Function Router gets event
+
+Picks runner to run it on and passes along event
+
+### Runner gets event
+
+Takes function URL and looks up metadata from function registry. 
+Executes the function
+Returns CloudEvent with response as “data” body, strips out all the extension stuff and replace with response specific info.
+
+```json
+{
+  "cloud-events-version": "0.1",
+  "event-id": "6480da1a-5028-4301-acc3-fbae628207b3",
+  "source": "http://example.com/repomanager",
+  "event-type": "http-request",
+  "event-type-version": "v1.5",
+  "event-time": "2018-04-01T23:12:34Z",
+  "content-type": "application/json",
+  "data": {
+     RESPONSE_BODY
+  },
+  “extensions”: {
+    “protocol”: {
+       “headers”: …
+       Etc, same as we do now for JSON format responses 
+    },
+  }
+}
+```
+
+## CLI changes
+
+### fn deploy
+
+func.yaml barely changes:
+
+```yaml
+function:
+  name: yodawg
+  image: hub.docker.com/fnproject/hello
+  version: 0.0.7
+  memory: 42
+
+trigger:
+  name: myApp/sayhello
+  type: http
+```
+
+trigger in yaml is optional. if trigger is not specified, it can be specified:
+
+`fn deploy --trigger-http sayhello`
+
+since namespace is not specified, this will create a function at:
+
+```
+/ns/_/funcs/yodawg
+```
+
+and a trigger at:
+
+```
+/ns/_/apps/myApp/triggers/sayhello
+{
+  name: sayhello
+  func: yodawg:0.0.7
+  type: http
+}
+```
+
+since the above `func: ` does not specify a host name, the func will be
+expected to exist on the fn service where this trigger exists. a full url is
+also possible, see below example.
+
+it is also possible to specify a `func.yaml` with a remote function, where the
+trigger will be deployed but the function in the working directory will not be
+built and pushed to a docker registry, say if we had a func:
+
+```
+trigger:
+  name: /sayhello
+  type: http
+  func: hub.fnproject.io/funcytown/hello
+```
+
+in all of the above examples, a user will end up with a route to call:
+
+`http://my.fn.com/sayhello`
+
+TODO do the above with namespace instead and don't have `_` magic?
+
+### CRUD
+
+we'll need to change crud around routes to have CRUD for functions and
+triggers, and add namespaces.
+
+## FDKs
+
+There are some decisions to make around added sugar we want to sprinkle on top
+of cloud events. We also may want to allow users to receive/output raw cloud-events
+themselves if they would like to, or at least have a way to set each of the
+fields and extensions (the catch being if users can define their own they may
+add additional fields) -- of course, it's possible to not use FDK and do this,
+so my thinking is that we make FDKs more to simplify things than anything.
+We can add features we'd like to FDKs as we go down the road. My thinking for
+base level FDK functionality is:
+
+FDKs will simply handle the json version of cloud-events defined here: 
+https://github.com/cloudevents/spec/blob/master/json-format.md
```

Comment:
there is a discussion to trigger a function with more than one event, how would we handle this case?

Reply:
Why would one triggered invocation of a function take two events? Can you explain a use case here?

Reply:
AFAIR the idea is to have a threshold for a function invocation, i.e. the action (function call) must happen only when N events appear. In real-world deployment function, most of the times get triggered by an Event Service. It capable of defining the threshold for a function invocation.

Another use case is butch event distribution (azure supports that, azure event grid is the closest one to cloudevent spec).

P.S.: can't find cloudevent doc where that discussion happened.

Reply:
Hmm well I mean we could handle batch events and fire them one at a time into the container other wise make the functions spec be that you consume a list of cloud events and output a list of them as well.

Reply:
Seems wierd / cumbersome to handle 1 or a list though.

Reply:
i guess this would be up to the event source to manage that threshold and fire when ready (not fn), so to that we can now say 'sure', but it would have to fire multiple discrete events still there's no batching. since we have a 1:1 mapping of an event to a container at the moment, it seems weird to start trying to add guarantees around running multiple events through the `/run` endpoint at this point.